# PROBLEM STATEMENT

# I have been hired as a data analyst by a sports management company. They are interested in forming a new team for the upcoming IPL Season 2018 and want your expertise to suggest players that will maximize their chances of winning matches. Your task is to analyze the IPL dataset and recommend the top-performing players in various positions to include in the new team.  

# DATA DESCRIPTION

# Using the "IPL_Ball_by_Ball_2008_2022"  dataset.It contains comprehensive data on IPL matches, teams, players, and performance metrics.

## Below is the data dictionary explaining the various columns of the dataset: ID, innings, overs, ballnumber, batter, bowler,non-striker, batsman_run, extras_run, total_run, non_boundary, isWicketDelivery, BattingTeam


## we need to form a IPL team from the given information


## total 11 players needed to form a team in which 4 players are in extra 

# importing required libraries 

In [18]:
import pandas as pd
import matplotlib.pyplot as plt 

# Reading the Dataset

In [19]:
df = pd.read_csv(r'C:\Users\user\Documents\learn and build\archive\IPL_Ball_by_Ball_2008_2022.csv')

In [20]:
df

ID  innings  overs  ballnumber       batter          bowler  \
0       1312200        1      0           1  YBK Jaiswal  Mohammed Shami   
1       1312200        1      0           2  YBK Jaiswal  Mohammed Shami   
2       1312200        1      0           3   JC Buttler  Mohammed Shami   
3       1312200        1      0           4  YBK Jaiswal  Mohammed Shami   
4       1312200        1      0           5  YBK Jaiswal  Mohammed Shami   
...         ...      ...    ...         ...          ...             ...   
225949   335982        2     14           5      P Kumar        I Sharma   
225950   335982        2     14           6     SB Joshi        I Sharma   
225951   335982        2     14           7      P Kumar        I Sharma   
225952   335982        2     15           1     SB Joshi       LR Shukla   
225953   335982        2     15           2     SB Joshi       LR Shukla   

        non-striker extra_type  batsman_run  extras_run  total_run  \
0        JC Buttler        NaN            0           0          0   
1        JC Buttler    legbyes            0           1          1   
2       YBK Jaiswal        NaN            1           0          1   
3        JC Buttler        NaN            0           0          0   
4        JC Buttler        NaN            0           0          0   
...             ...        ...          ...         ...        ...   
225949     SB Joshi    legbyes            0           1          1   
225950      P Kumar        NaN            1           0          1   
225951     SB Joshi        NaN            0           0          0   
225952      P Kumar      wides            0           1          1   
225953      P Kumar        NaN            0           0          0   

        non_boundary  isWicketDelivery player_out    kind fielders_involved  \
0                  0                 0        NaN     NaN               NaN   
1                  0                 0        NaN     NaN               NaN   
2                  0                 0        NaN     NaN               NaN   
3                  0                 0        NaN     NaN               NaN   
4                  0                 0        NaN     NaN               NaN   
...              ...               ...        ...     ...               ...   
225949             0                 0        NaN     NaN               NaN   
225950             0                 0        NaN     NaN               NaN   
225951             0                 0        NaN     NaN               NaN   
225952             0                 0        NaN     NaN               NaN   
225953             0                 1   SB Joshi  caught       BB McCullum   

                        BattingTeam  
0                  Rajasthan Royals  
1                  Rajasthan Royals  
2                  Rajasthan Royals  
3                  Rajasthan Royals  
4                  Rajasthan Royals  
...                             ...  
225949  Royal Challengers Bangalore  
225950  Royal Challengers Bangalore  
225951  Royal Challengers Bangalore  
225952  Royal Challengers Bangalore  
225953  Royal Challengers Bangalore  

[225954 rows x 17 columns]

In [21]:
df.shape

(225954, 17)

# printing top and botom rows to check the data entered is in correct manner

In [22]:
df.head()

ID  innings  overs  ballnumber       batter          bowler  \
0  1312200        1      0           1  YBK Jaiswal  Mohammed Shami   
1  1312200        1      0           2  YBK Jaiswal  Mohammed Shami   
2  1312200        1      0           3   JC Buttler  Mohammed Shami   
3  1312200        1      0           4  YBK Jaiswal  Mohammed Shami   
4  1312200        1      0           5  YBK Jaiswal  Mohammed Shami   

   non-striker extra_type  batsman_run  extras_run  total_run  non_boundary  \
0   JC Buttler        NaN            0           0          0             0   
1   JC Buttler    legbyes            0           1          1             0   
2  YBK Jaiswal        NaN            1           0          1             0   
3   JC Buttler        NaN            0           0          0             0   
4   JC Buttler        NaN            0           0          0             0   

   isWicketDelivery player_out kind fielders_involved       BattingTeam  
0                 0        NaN  NaN               NaN  Rajasthan Royals  
1                 0        NaN  NaN               NaN  Rajasthan Royals  
2                 0        NaN  NaN               NaN  Rajasthan Royals  
3                 0        NaN  NaN               NaN  Rajasthan Royals  
4                 0        NaN  NaN               NaN  Rajasthan Royals

In [23]:
df.tail()

ID  innings  overs  ballnumber    batter     bowler non-striker  \
225949  335982        2     14           5   P Kumar   I Sharma    SB Joshi   
225950  335982        2     14           6  SB Joshi   I Sharma     P Kumar   
225951  335982        2     14           7   P Kumar   I Sharma    SB Joshi   
225952  335982        2     15           1  SB Joshi  LR Shukla     P Kumar   
225953  335982        2     15           2  SB Joshi  LR Shukla     P Kumar   

       extra_type  batsman_run  extras_run  total_run  non_boundary  \
225949    legbyes            0           1          1             0   
225950        NaN            1           0          1             0   
225951        NaN            0           0          0             0   
225952      wides            0           1          1             0   
225953        NaN            0           0          0             0   

        isWicketDelivery player_out    kind fielders_involved  \
225949                 0        NaN     NaN               NaN   
225950                 0        NaN     NaN               NaN   
225951                 0        NaN     NaN               NaN   
225952                 0        NaN     NaN               NaN   
225953                 1   SB Joshi  caught       BB McCullum   

                        BattingTeam  
225949  Royal Challengers Bangalore  
225950  Royal Challengers Bangalore  
225951  Royal Challengers Bangalore  
225952  Royal Challengers Bangalore  
225953  Royal Challengers Bangalore

In [24]:
df.describe()

ID        innings          overs     ballnumber  \
count  2.259540e+05  225954.000000  225954.000000  225954.000000   
mean   8.320470e+05       1.483868       9.185679       3.619750   
std    3.379542e+05       0.503104       5.681797       1.810633   
min    3.359820e+05       1.000000       0.000000       1.000000   
25%    5.012620e+05       1.000000       4.000000       2.000000   
50%    8.297370e+05       1.000000       9.000000       4.000000   
75%    1.178395e+06       2.000000      14.000000       5.000000   
max    1.312200e+06       6.000000      19.000000      10.000000   

         batsman_run     extras_run     total_run   non_boundary  \
count  225954.000000  225954.000000  225954.00000  225954.000000   
mean        1.243523       0.066907       1.31043       0.000093   
std         1.618166       0.341470       1.60605       0.009640   
min         0.000000       0.000000       0.00000       0.000000   
25%         0.000000       0.000000       0.00000       0.000000   
50%         1.000000       0.000000       1.00000       0.000000   
75%         1.000000       0.000000       1.00000       0.000000   
max         6.000000       7.000000       7.00000       1.000000   

       isWicketDelivery  
count     225954.000000  
mean           0.049351  
std            0.216600  
min            0.000000  
25%            0.000000  
50%            0.000000  
75%            0.000000  
max            1.000000

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 225954 entries, 0 to 225953
Data columns (total 17 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   ID                 225954 non-null  int64 
 1   innings            225954 non-null  int64 
 2   overs              225954 non-null  int64 
 3   ballnumber         225954 non-null  int64 
 4   batter             225954 non-null  object
 5   bowler             225954 non-null  object
 6   non-striker        225954 non-null  object
 7   extra_type         12049 non-null   object
 8   batsman_run        225954 non-null  int64 
 9   extras_run         225954 non-null  int64 
 10  total_run          225954 non-null  int64 
 11  non_boundary       225954 non-null  int64 
 12  isWicketDelivery   225954 non-null  int64 
 13  player_out         11151 non-null   object
 14  kind               11151 non-null   object
 15  fielders_involved  7988 non-null    object
 16  BattingTeam        2

In [26]:
df.nunique()

ID                   950
innings                6
overs                 20
ballnumber            10
batter               605
bowler               472
non-striker          595
extra_type             5
batsman_run            7
extras_run             7
total_run              8
non_boundary           2
isWicketDelivery       2
player_out           573
kind                  10
fielders_involved    535
BattingTeam           18
dtype: int64

# at first we will do cleaning our data
### it includes treating null values

In [27]:
df.isnull().sum()

ID                        0
innings                   0
overs                     0
ballnumber                0
batter                    0
bowler                    0
non-striker               0
extra_type           213905
batsman_run               0
extras_run                0
total_run                 0
non_boundary              0
isWicketDelivery          0
player_out           214803
kind                 214803
fielders_involved    217966
BattingTeam               0
dtype: int64

### percentage of not available data

In [28]:
df['extra_type'].isnull().mean()*100

94.66749869442452

In [29]:
df['player_out'].isnull().mean()*100

95.06492471919063

In [30]:
df['kind'].isnull().mean()*100

95.06492471919063

In [31]:
df['fielders_involved'].isnull().mean()*100

96.4647671649982

### But the given data in these column may be helpfull in selecting players so we do not delete these columns

## Some of the columns is not relevant like ID and innings so we delete them

In [32]:
df.drop(['ID', 'innings'], axis = 1, inplace = True)

In [33]:
df.columns

Index(['overs', 'ballnumber', 'batter', 'bowler', 'non-striker', 'extra_type',
       'batsman_run', 'extras_run', 'total_run', 'non_boundary',
       'isWicketDelivery', 'player_out', 'kind', 'fielders_involved',
       'BattingTeam'],
      dtype='object')

# In player_out column, nan is replaced by not out as when the player is out the data is mentioned
## nan of extra_type and kind will be filled with no
## nan of fielders_involved will be filled with none 

In [34]:
df['player_out'].fillna('Not Out', inplace = True)

In [35]:
df['extra_type'].fillna('No', inplace = True)

In [36]:
df['fielders_involved'].fillna('None', inplace = True)

In [37]:
df

overs  ballnumber       batter          bowler  non-striker  \
0           0           1  YBK Jaiswal  Mohammed Shami   JC Buttler   
1           0           2  YBK Jaiswal  Mohammed Shami   JC Buttler   
2           0           3   JC Buttler  Mohammed Shami  YBK Jaiswal   
3           0           4  YBK Jaiswal  Mohammed Shami   JC Buttler   
4           0           5  YBK Jaiswal  Mohammed Shami   JC Buttler   
...       ...         ...          ...             ...          ...   
225949     14           5      P Kumar        I Sharma     SB Joshi   
225950     14           6     SB Joshi        I Sharma      P Kumar   
225951     14           7      P Kumar        I Sharma     SB Joshi   
225952     15           1     SB Joshi       LR Shukla      P Kumar   
225953     15           2     SB Joshi       LR Shukla      P Kumar   

       extra_type  batsman_run  extras_run  total_run  non_boundary  \
0              No            0           0          0             0   
1         legbyes            0           1          1             0   
2              No            1           0          1             0   
3              No            0           0          0             0   
4              No            0           0          0             0   
...           ...          ...         ...        ...           ...   
225949    legbyes            0           1          1             0   
225950         No            1           0          1             0   
225951         No            0           0          0             0   
225952      wides            0           1          1             0   
225953         No            0           0          0             0   

        isWicketDelivery player_out    kind fielders_involved  \
0                      0    Not Out     NaN              None   
1                      0    Not Out     NaN              None   
2                      0    Not Out     NaN              None   
3                      0    Not Out     NaN              None   
4                      0    Not Out     NaN              None   
...                  ...        ...     ...               ...   
225949                 0    Not Out     NaN              None   
225950                 0    Not Out     NaN              None   
225951                 0    Not Out     NaN              None   
225952                 0    Not Out     NaN              None   
225953                 1   SB Joshi  caught       BB McCullum   

                        BattingTeam  
0                  Rajasthan Royals  
1                  Rajasthan Royals  
2                  Rajasthan Royals  
3                  Rajasthan Royals  
4                  Rajasthan Royals  
...                             ...  
225949  Royal Challengers Bangalore  
225950  Royal Challengers Bangalore  
225951  Royal Challengers Bangalore  
225952  Royal Challengers Bangalore  
225953  Royal Challengers Bangalore  

[225954 rows x 15 columns]

## we have cleaned our data.Now we do Exploratory Data Analysis and We will explore the dataset 

## players according to their mean of runs scored 

In [38]:
batters_count2 = df.groupby('batter').agg({'ballnumber': 'count', 'batsman_run': 'mean'})
batters_count2.rename(columns={"batsman_run":"batsman_run_mean"})

ballnumber  batsman_run_mean
batter                                      
A Ashish Reddy         196          1.428571
A Badoni               139          1.158273
A Chandila               7          0.571429
A Chopra                75          0.706667
A Choudhary             20          1.250000
...                    ...               ...
Yash Dayal               1          0.000000
Yashpal Singh           67          0.701493
Younis Khan              7          0.428571
Yuvraj Singh          2207          1.247848
Z Khan                 141          0.829787

[605 rows x 2 columns]

In [39]:
no_of_runs = df.groupby('batter').agg({'batsman_run': 'sum', 'batter':'count'})
no_of_runs.rename(columns={'batter':'no_of_balls'}, inplace = True)
no_of_runs

batsman_run  no_of_balls
batter                                  
A Ashish Reddy          280          196
A Badoni                161          139
A Chandila                4            7
A Chopra                 53           75
A Choudhary              25           20
...                     ...          ...
Yash Dayal                0            1
Yashpal Singh            47           67
Younis Khan               3            7
Yuvraj Singh           2754         2207
Z Khan                  117          141

[605 rows x 2 columns]

# No. of times a Batsman get out 

In [40]:
a=df.groupby('batter').agg({'batsman_run':'sum', 'isWicketDelivery':'sum'})
a.rename(columns={'isWicketDelivery': 'out_no'}, inplace = True)
a

batsman_run  out_no
batter                             
A Ashish Reddy          280      15
A Badoni                161       9
A Chandila                4       1
A Chopra                 53       5
A Choudhary              25       2
...                     ...     ...
Yash Dayal                0       1
Yashpal Singh            47       4
Younis Khan               3       1
Yuvraj Singh           2754     109
Z Khan                  117      15

[605 rows x 2 columns]

# Player who scored maximum times 6 

In [41]:
df[df['batsman_run'] == 6]['batter'].describe()

count        10666
unique         400
top       CH Gayle
freq           359
Name: batter, dtype: object

### total 10666 times players scored 6 , in which  400 unique players scored 6 runs and CH Gayle scored 6 runs 359 times

# Top 5 players who scored maximum times 6

In [42]:
top_5_batsManwith6= df[df['batsman_run']==6]['batter'].value_counts()[df[df['batsman_run']==6]['batter'].value_counts() >= 223]
top_5_batsManwith6

CH Gayle          359
AB de Villiers    253
RG Sharma         241
MS Dhoni          229
KA Pollard        224
Name: batter, dtype: int64

# Player who scored maximum times 4

In [43]:
df[df['batsman_run'] == 4]['batter'].describe()

count        25500
unique         484
top       S Dhawan
freq           701
Name: batter, dtype: object

# Top 5 players who scored maximum times 4

In [44]:
top_5_batsManwith4= df[df['batsman_run']==4]['batter'].value_counts()[df[df['batsman_run']==4]['batter'].value_counts() >= 500]
top_5_batsManwith4

S Dhawan     701
V Kohli      581
DA Warner    577
RG Sharma    519
SK Raina     506
Name: batter, dtype: int64

### Highest number of runs scored by V Kohli followed by other players

## Now looking for the players who get out 

In [45]:
df.player_out.value_counts()[1:11]

RG Sharma     194
V Kohli       183
RV Uthappa    180
S Dhawan      179
SK Raina      171
KD Karthik    163
AT Rayudu     145
DA Warner     142
G Gambhir     136
AM Rahane     132
Name: player_out, dtype: int64

# Calculate strike rate for finding the chances of hitting

In [46]:
no_of_runs['strike_rate'] = no_of_runs['batsman_run']/no_of_runs['no_of_balls']*100
no_of_runs

batsman_run  no_of_balls  strike_rate
batter                                               
A Ashish Reddy          280          196   142.857143
A Badoni                161          139   115.827338
A Chandila                4            7    57.142857
A Chopra                 53           75    70.666667
A Choudhary              25           20   125.000000
...                     ...          ...          ...
Yash Dayal                0            1     0.000000
Yashpal Singh            47           67    70.149254
Younis Khan               3            7    42.857143
Yuvraj Singh           2754         2207   124.784776
Z Khan                  117          141    82.978723

[605 rows x 3 columns]

In [47]:
no_of_runs.loc[no_of_runs['no_of_balls']>=600, 'best_strikerate'] =no_of_runs['strike_rate']
top5= no_of_runs.sort_values(by='best_strikerate',ascending=False).head(20)
top5

batsman_run  no_of_balls  strike_rate  best_strikerate
batter                                                                
AD Russell             2039         1212   168.234323       168.234323
SP Narine              1025          664   154.367470       154.367470
V Sehwag               2728         1833   148.827059       148.827059
AB de Villiers         5181         3487   148.580442       148.580442
GJ Maxwell             2320         1571   147.676639       147.676639
JC Buttler             2832         1955   144.859335       144.859335
N Pooran                912          630   144.761905       144.761905
PP Shaw                1588         1106   143.580470       143.580470
RR Pant                2851         2000   142.550000       142.550000
CH Gayle               4997         3516   142.121729       142.121729
KA Pollard             3437         2447   140.457703       140.457703
HH Pandya              1972         1406   140.256046       140.256046
MM Ali                  910          652   139.570552       139.570552
JM Bairstow            1291          935   138.074866       138.074866
YK Pathan              3222         2334   138.046272       138.046272
RA Tripathi            1798         1312   137.042683       137.042683
JA Morkel               975          712   136.938202       136.938202
CA Lynn                1329          974   136.447639       136.447639
DA Warner              5883         4313   136.401577       136.401577
SR Watson              3880         2892   134.163209       134.163209

### Top players strike rate 

# *Now analyse the data for Bowlers

In [48]:
df['bowler'].describe()

count       225954
unique         472
top       R Ashwin
freq          4024
Name: bowler, dtype: object

### Total 472 number of bowlers in which R Ashwin bowls 4024 times

In [49]:
no_of_wickets = df.groupby('bowler').agg({'ballnumber': 'count','total_run':'sum'})
no_of_wickets

ballnumber  total_run
bowler                               
A Ashish Reddy         270        400
A Badoni                12         12
A Chandila             234        245
A Choudhary            108        144
A Dananjaya             25         47
...                    ...        ...
YK Pathan             1184       1443
YS Chahal             2940       3699
Yash Dayal             205        296
Yuvraj Singh           882       1091
Z Khan                2276       2860

[472 rows x 2 columns]

## Players acoording to the number of wickets they take

In [50]:
no_of_wickets = df.groupby('bowler').agg({'isWicketDelivery': 'sum'})

In [51]:
no_of_wickets

isWicketDelivery
bowler                          
A Ashish Reddy                19
A Badoni                       2
A Chandila                    11
A Choudhary                    5
A Dananjaya                    0
...                          ...
YK Pathan                     46
YS Chahal                    172
Yash Dayal                    13
Yuvraj Singh                  39
Z Khan                       119

[472 rows x 1 columns]

# Now analysing the bowling rate 

In [52]:
no_of_wickets= df.groupby('bowler').agg({'total_run':'sum','bowler':'count','isWicketDelivery':'sum'})
no_of_wickets.rename(columns={'bowler':'no_of_balls'},inplace=True)
no_of_wickets

total_run  no_of_balls  isWicketDelivery
bowler                                                  
A Ashish Reddy        400          270                19
A Badoni               12           12                 2
A Chandila            245          234                11
A Choudhary           144          108                 5
A Dananjaya            47           25                 0
...                   ...          ...               ...
YK Pathan            1443         1184                46
YS Chahal            3699         2940               172
Yash Dayal            296          205                13
Yuvraj Singh         1091          882                39
Z Khan               2860         2276               119

[472 rows x 3 columns]

In [53]:
no_of_wickets['bowlingRate'] = no_of_wickets['total_run']/no_of_wickets['no_of_balls']*6
no_of_wickets

total_run  no_of_balls  isWicketDelivery  bowlingRate
bowler                                                               
A Ashish Reddy        400          270                19     8.888889
A Badoni               12           12                 2     6.000000
A Chandila            245          234                11     6.282051
A Choudhary           144          108                 5     8.000000
A Dananjaya            47           25                 0    11.280000
...                   ...          ...               ...          ...
YK Pathan            1443         1184                46     7.312500
YS Chahal            3699         2940               172     7.548980
Yash Dayal            296          205                13     8.663415
Yuvraj Singh         1091          882                39     7.421769
Z Khan               2860         2276               119     7.539543

[472 rows x 4 columns]

## Finding top players according to bowling rate

In [54]:
no_of_wickets.loc[no_of_wickets['no_of_balls'] >= 60, 'best_wicket_delivery'] =no_of_wickets['bowlingRate']
t= no_of_wickets.sort_values(by='best_wicket_delivery').head(10)
t

total_run  no_of_balls  isWicketDelivery  bowlingRate  \
bowler                                                                   
Mohsin Khan            202          206                14     5.883495   
Sohail Tanvir          275          265                24     6.226415   
A Chandila             245          234                11     6.282051   
FH Edwards             160          150                 6     6.400000   
Rashid Khan           2411         2232               119     6.481183   
JW Hastings             66           61                 3     6.491803   
SMSM Senanayake        211          195                 9     6.492308   
MJ Clarke               72           66                 3     6.545455   
SM Pollock             307          280                13     6.578571   
SM Harwood              74           67                 3     6.626866   

                 best_wicket_delivery  
bowler                                 
Mohsin Khan                  5.883495  
Sohail Tanvir                6.226415  
A Chandila                   6.282051  
FH Edwards                   6.400000  
Rashid Khan                  6.481183  
JW Hastings                  6.491803  
SMSM Senanayake              6.492308  
MJ Clarke                    6.545455  
SM Pollock                   6.578571  
SM Harwood                   6.626866

In [55]:
df['kind'].value_counts()


caught                   6837
bowled                   1944
run out                  1007
lbw                       685
stumped                   325
caught and bowled         323
hit wicket                 14
retired hurt               13
obstructing the field       2
retired out                 1
Name: kind, dtype: int64

## we can find wicket keeper by kind = stumped 

In [56]:
df.loc[df['kind'] == 'stumped', 'wicket_out'] =df['fielders_involved']
wicket_keeper= df.groupby('fielders_involved').agg({'wicket_out':'count'})
wicket_keeper.sort_values(by='wicket_out',ascending=False).head(5)

wicket_out
fielders_involved            
MS Dhoni                   39
KD Karthik                 34
RV Uthappa                 32
WP Saha                    22
RR Pant                    18

### best wicket keeper is MS Dhoni

In [57]:
df.loc[df['kind'] == 'caught', 'catchout'] =df['fielders_involved']
catchout= df.groupby('fielders_involved').agg({'catchout':'count'})
catchout.sort_values(by='catchout',ascending=False).head(5)

catchout
fielders_involved          
MS Dhoni                135
KD Karthik              133
AB de Villiers          120
SK Raina                106
KA Pollard               97

## common players that do both bowling and batting

In [58]:
df[df['bowler']==df['batter']]['bowler'].any()

False

### no such player that do both batting and bowler

# CONCLUSION:


## Batsman: 

# Batsman choosen:
### TH David
### LJ Wright
### S Dhawan 
### CH Gayle

# Bowlers choosen:
### Mohsin Khan	
### Sohail Tanvir	 
### A Chandila
### Rashid Khan

# Wicket Keeper:
## MS Dhoni

# Fielders:
## KD Karthik
## AB de Villiers